solo-io/gloo
# Istio X Forwarded Host toggle

# Description

Expose a `global.istioIntegration.AppendXForwardedHost` Helm value which allows users opt out of setting the `x-forwarded-host` header on requests using the Istio plugin

## API changes
- Added `IstioOptions` message to the `Settings` proto with a `AppendXForwardedHost` boolean field

## Code changes
- Istio Integration plugin gets a new `appendXForwardedHost` field which gets its value from the `Gloo.IstioOptions.AppendXForwardedHost` field on `Init()`
  - The output `routeAction` only has `Route.AppendXForwardedHost` set to `true` if the plugin's corresponding field is `true`
  - Add unit tests for this behavior

## Helm changes
- Expose `global.IstioIntegration.AppendXForwardedHost` Helm value
  - This value is propagated directly as `gloo.IstioOptions.AppendXForwardedHeader` in the Settings template
  - This value is set to `true` in `values-template.yaml`, making `true` the default
  - Add helm test for this behavior

# Context

This replaces #9041

## Interesting decisions
 
The location of the value to config was chosen to make the most sense in the immediate context
For Helm it made the most sense to add the Istio X Forwarded Host valeu to the existing `IstioIntegration` section
For the Settings proto we follow the pattern of creating messages for plugin-specific config under the `Gloo` message
Users are expected to configure via Helm, so the fact that the location of the value doesn't match the Settings CR 1:1 is not expected to lead to significant confustion
Both are included in reference docs

## Testing steps

I have manually verified the intended behavior using the following steps:
- Use `./ci/kind/setup-kind.sh` to setup a KinD cluster
- Run `make build-test-chart` to create a local Helm chart
- Follow the [docs guide for setting up Edge with Istio](https://docs.solo.io/gloo-edge/latest/guides/integrations/service_mesh/istio/), but using the local chart
  - The installation command becomes
  ```
  helm install gloo ./install/helm/gloo --namespace gloo-system --create-namespace -f value-overrides.yaml
  ```
- Observe that the Settings CR has `gloo.istioOptions.appendXForwardedHost` set to `true`
- At step 3 of the "Set up and verify the mTLS connection" section of the guide, observe that the response contains the `X-Forwarded-Host` header:
```
{
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.default",
    "User-Agent": "curl/8.1.2",
    "X-B3-Sampled": "0",
    "X-B3-Spanid": "c2f0a67a569f15dd",
    "X-B3-Traceid": "1694608ef69f94e4c2f0a67a569f15dd",
    "X-Forwarded-Host": "www.example.com"
  }
}
```
- Modify `value-overrides.yaml` to add `global.istioIntegration.XForwardedHost=false`
- Upgrade the helm installation:
```
helm upgrade gloo ./install/helm/gloo --namespace gloo-system --create-namespace -f value-overrides.yaml
```
- Observe that the Settings CR has `gloo.istioOptions.appendXForwardedHost` set to `false`
- Execute the same `curl` from the guide and observe that the response does not contain the `X-Forwarded-Host` header:
```
{
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.default",
    "User-Agent": "curl/8.1.2",
    "X-B3-Sampled": "0",
    "X-B3-Spanid": "0f74a457a0753ca9",
    "X-B3-Traceid": "0bf1353677a9eaeb0f74a457a0753ca9"
  }
}
```

Further testing and validation is expected to be customer success' responsibility

# Checklist:
